File: clients/soa/python/src/crosslab_soa_client/service.py
from abc import ABC, abstractmethod
from typing import Literal


class Service(ABC):
    service_type: str
    service_direction: Literal["in", "out", "inout"]
    service_id: str

    @abstractmethod
    def getMeta(self) -> dict:
        return {
            "serviceId": self.service_id,
            "serviceType": self.service_type,
            "serviceDirection": self.service_direction,
        }

    @abstractmethod
    def setupConnection(self, connection, serviceConfig):
        pass

File: clients/soa/js/src/service.ts
import { ServiceConfiguration } from './deviceMessages.js';
import { PeerConnection } from './peer/connection.js';

export type ServiceDirection = 'consumer' | 'producer' | 'prosumer';
export const Consumer: ServiceDirection = 'consumer';
export const Producer: ServiceDirection = 'producer';
export const Prosumer: ServiceDirection = 'prosumer';
export { ServiceConfiguration };

export interface Service<T extends string = string, ID extends string=string, FS extends Record<string, any> = Record<string, any>> {
  getMeta: () => {
    serviceId: ID;
    serviceType: T;
    serviceDirection: ServiceDirection;
  };
  serviceType: T;
  serviceId: ID;
  serviceDirection: ServiceDirection;
  setupConnection(connection: PeerConnection, serviceConfig: ServiceConfiguration): void;
  get_state(): FS
}
